blockchain/blockdao/blockdao.go

```diff
@@ -81,48 +80,59 @@ var (
 	ErrNotOpened = errors.New("DB is not opened")
 )
 
-// BlockDAO represents the block data access object
-type BlockDAO interface {
-	Start(ctx context.Context) error
-	Stop(ctx context.Context) error
-	Commit() error
-	GetBlockHash(uint64) (hash.Hash256, error)
-	GetBlockHeight(hash.Hash256) (uint64, error)
-	GetBlock(hash.Hash256) (*block.Block, error)
-	GetBlockByHeight(uint64) (*block.Block, error)
-	GetTipHeight() (uint64, error)
-	GetTipHash() (hash.Hash256, error)
-	Header(hash.Hash256) (*block.Header, error)
-	Body(hash.Hash256) (*block.Body, error)
-	Footer(hash.Hash256) (*block.Footer, error)
-	GetActionByActionHash(hash.Hash256, uint64) (action.SealedEnvelope, error)
-	GetReceiptByActionHash(hash.Hash256, uint64) (*action.Receipt, error)
-	GetReceipts(uint64) ([]*action.Receipt, error)
-	PutBlock(*block.Block) error
-	DeleteTipBlock() error
-	IndexFile(uint64, []byte) error
-	GetFileIndex(uint64) ([]byte, error)
-	KVStore() db.KVStore
-}
-
-type blockDAO struct {
-	compressBlock bool
-	kvstore       db.KVStore
-	indexer       blockindex.Indexer
-	htf           db.RangeIndex
-	kvstores      sync.Map //store like map[index]db.KVStore,index from 1...N
-	topIndex      atomic.Value
-	timerFactory  *prometheustimer.TimerFactory
-	lifecycle     lifecycle.Lifecycle
-	headerCache   *cache.ThreadSafeLruCache
-	bodyCache     *cache.ThreadSafeLruCache
-	footerCache   *cache.ThreadSafeLruCache
-	cfg           config.DB
-	mutex         sync.RWMutex // for create new db file
-}
+type (
+	// BlockDAO represents the block data access object
+	BlockDAO interface {
+		Start(ctx context.Context) error
+		Stop(ctx context.Context) error
+		GetBlockHash(uint64) (hash.Hash256, error)
+		GetBlockHeight(hash.Hash256) (uint64, error)
+		GetBlock(hash.Hash256) (*block.Block, error)
+		GetBlockByHeight(uint64) (*block.Block, error)
+		GetTipHeight() (uint64, error)
+		GetTipHash() (hash.Hash256, error)
+		Header(hash.Hash256) (*block.Header, error)
+		Body(hash.Hash256) (*block.Body, error)
+		Footer(hash.Hash256) (*block.Footer, error)
+		GetActionByActionHash(hash.Hash256, uint64) (action.SealedEnvelope, error)
+		GetReceiptByActionHash(hash.Hash256, uint64) (*action.Receipt, error)
+		GetReceipts(uint64) ([]*action.Receipt, error)
+		PutBlock(*block.Block) error
+		Commit() error
+		DeleteTipBlock() error
+		IndexFile(uint64, []byte) error
+		GetFileIndex(uint64) ([]byte, error)
+		KVStore() db.KVStore
+	}
+
+	// BlockIndexer defines an interface to accept block to build index
+	BlockIndexer interface {
```

Comment:
if i understand correctly, the intention is that the indexer in BlockDAO only needs to have these 5 API.

would suggest name this interface to like "IndexPutter" and still define it under /blockindex, define another BlockIndexer interface under BlockDAO is kind of confusing

Reply:
Naming doesn't matter. For example, each package could define their own blockchain interfaces. Moreover, defining this in blockdao is on purpose. We discussed this problem before.